### ISSUE TITLE
Discussion

### ISSUE THREAD

**Issue by helfer** [apollostack/graphql-decorators#1]
Discuss the GraphQL schema decorator proposal here.

**Comment by almilo** [comment]
Hi @helfer I am super interested in participating here. I started implementing a PoC on my own some time ago (see https://github.com/graphql/graphql-js/pull/265#issuecomment-189948879) and had thought to update some parts of it with your cool approach from graphql-tools on how to decompose and build executable schemas but this repo makes it redundant so I would prefer to contribute here.
I would like first to agree on a list of possible decorators here and then we can discuss who works on which one.
what do you think?

**Comment by helfer** [comment]
@almilo Yes, that sounds great! I saw the example you already built with the Swiss trains and I really liked it. Let's make a list of decorators and then discuss the design in a separate issue for each resolver before we start implementing. Some of them may be trivial and won't need much discussion, but for others it could be more complicated.

Off the top of my head, here are a few decorators to start with:

- description (for adding descriptions)
- deprecation (for adding deprecation reason)
- validation (for adding validations to arguments or fields)
- logging or profiling
- authorization (requiring permissions for accessing types or fields)
Please add more and feel free to open issues for discussing the implementation of some of them. I think particularly the rest decorator is tricky. We should make sure the interface covers most use cases  (including forwarding auth tokens) but is still easy to use in most cases.

**Comment by helfer** [comment]
@almilo There's also still a bit of work to do to implement the decorator parsing and application to the schema. Right now I have examples for types and fields with GraphQL-JS, but none so far with the schema language. I'm blocked on the PR https://github.com/graphql/graphql-js/pull/371 for GraphQL-JS to make it work with Schema and Args as well.

We could already start implementing the parser for decorators, but we would have to wait for the same PR to be merged (and a new version of GraphQL-JS to be published) to make it work with exrpress-graphql or other tools that use GraphQL-JS.

**Comment by almilo** [comment]
@helfer which is your strategy to implement the parser? are you thinking of creating a new parser? if so, which tool are you thinking of (hand-coded, ANTLR,...)? or are you maybe planning on forking the existing one in graphql-js to extend it?

**Comment by almilo** [comment]
@helfer have you also thought of an scenario where you have an schema without decorators:
```
// schema.js
type Person {
  uuid: String!
  name: String!
  longName: String
  friends: [Person]
}
```
 and you want to extend it with the decorations by merging / importing a second file with a partial schema?:
```
// extension.js
+connector(storage: "mongoDB")
+id(fields: ["uuid"])
type Person {
}
```
This would allow to import an existing client-side schema in shorthand format and, without modifying the shorthand file, just add the required decorators with a second file.

**Comment by helfer** [comment]
I definitely want to fork the current parser, the idea being that eventually this should make its way back into GraphQL-JS.

Regarding defining the decorators in a second file, do you think that will be a common requirement? I'm thinking that the decorators should be present on the client schema as well so having them in a separate file would be unnecessary, but I could see its use in making schemas composable (for example by adding a new field to an existing type in a schema, or by adding additional decorators).

Not sure though if having decorators in a separate file makes the schema harder to understand. If you really want to add extra decorators, you could still always add them in the code (I'm thinking of adding  a method called `addDecorator` for that.

**Comment by OlegIlyenko** [comment]
@helfer as you asked, here is my feedback :)

Thanks a lot for putting effort in this work! I think it's very interesting idea, in fact I implemented something similar in sangria as well. Details may be a bit different, but idea is similar. 

I definitely see need for user-defined schema meta-information in some form (annotations for instance). Directives have different purpose and mostly provided for a client to "configure" an execution engine. Annotations on the other hand provide a way for a server to communicate additional structured information to a client (for instance, whether some field requires authorization, and if yes, then which permissions are required to access it).

I see decorators shown in the proposal a bit differently though. They combine meta-information with an instruction to do something with field/type (decorate it in some way). There are a few concerns I have with this approach:

First is separations of concerns. I think it would be helpful to treat these 2 concepts independently: annotations (exposed meta-information) and decorators with additional information and logic for a server implementor. One of the reasons for this is that it's often not easy to express all necessary in isolation: I many cases, which I faced before, it's necessary to make full query analysis in order to implement a decorator. 

Let's take authorization as an example. As I mentioned before, I implemented something similar in sangria: middleware and query reducers. There is also things called `FieldTag`s which one can add to a filed in order to provide meta-information - they on themselves do not have any code or logic (so they a similar to annotations). One can create a field tag like `PermissionRequired("VIEW_PRODUCT")`, for instance.

Middleware object will contain all of the logic for authorization and would be called for every single field in the query. It becomes much more flexible solution, if you have one instance of middleware for the whole query and it is able to see all of the fields (in spirit of a visitor pattern). In case of auth, it can apply some default rules to the whole schema, for example, or apply some rules to a hierarchy of selected fields. In case of caching middleware, it can hold instances of caches for the whole query (something that you normally don't really want to put in `rootValue` or recently added `contextValue`).

Another example, where it is useful to have one instance of middleware for a query, is query instrumentation. For example in our production setup we report execution time of every single GraphQL field to InfluxDB. We also have counters for deprecated field usage. You can't really cover this use-case with decorators because you need to manually decorate every single field and type in the schema. This is also an example of instrumentation logic which does not require or expose any kind of annotations/meta-information.

In my experience it is very helpful and flexible to separate meta-information (in form of annotations) and runtime component in form of middleware.

Where it becomes even more helpful is query reducers, or in other words static query analyzers. I already faced several use cases where it becomes really helpful. Let's look at these 3 exemples:

1. Coming back to auth example. If you have partially protected schema (some fields are accessible without auth), then it is very helpful to know beforehand whether query contains any protected fields. Normally it takes a while to authenticate user and fetch auth information (you either need to do to the DB or make an HTTP call to an auth service or something like this), so ideally you want to prevent this if it is not necessary. This is where query reducers come into play. They can collect annotations from the query and then make decision whether it is necessary to fetch user auth info or not. Again, in this case it's not enough to to have local logic only in the decorator. In should be able to make judgment based on the whole query, but still it needs meta-additional information on a field level. (this is actually a real use-case, you can find discussion here: https://github.com/sangria-graphql/sangria/issues/88)

2. Query complexity/cost analysis. Not all fields are equal - some go to the database and fetch 100s of object, others just return a String value from a memory. If you would like to estimate the query costs without executing it, you need to collect costs for all of the fields in the query and then sum them. In this case individual fields may customize their complexity (possibly via annotations), but the logic (analyzer) should be able to aggregate this information for all of the fields.

3. From readme: " For instance, the +validateRange decorator modifies the behavior of the containing field's resolve function.". I would argue that this is suboptimal semantics for argument validation decorators. Argument values are already known even before query gets executed. In my opinion, it would be undesirable, for instance, to make very expensive DB query (or HTTP call) to fetch 100s of objects and load them in memory just to find out that field argument somewhere down the line is malformed. It would help if validation logic is performed even before query is executed, for instance, with the help of query reducers. But then I can't really call it a decorator anymore.

Also if you would separate annotation and middleware logic it becomes super easy to replace middleware implementation, or disable it. For instance if you would settle at some point on a particular set of security primitives and share them in form annotations, them different security/auth libraries can provide their own implementations of middleware which supports these shared annotations.

uh, it was a lot of text :) Hope this helps and contributes a bit to this discussion. I just would like to point out that here I expressed my point of view and understanding of the subject, which is not set in stone. If you see that I misunderstood something or you disagree with some points, then please let me know, I would appreciate it (and would be happy to discuss further).

**Comment by helfer** [comment]
@OlegIlyenko Thanks a lot for the great feedback!

> I think it would be helpful to treat these 2 concepts independently: annotations (exposed meta-information) and decorators with additional information and logic for a server implementor.
Yeah, that makes perfect sense. I think @freiksenet made the same point in another issue and I think you guys convinced me that it's better to separate the two.

In fact, Lee Byron's recent PR https://github.com/graphql/graphql-js/pull/376 implements schema directives without semantic meaning, so I think I'll definitely rework the proposal based on that. I think they cover the part about decorators that was very important, which is that there is some validation for the arguments, and that directives have to be declared, and cannot just be randomly added.

> I many cases, which I faced before, it's necessary to make full query analysis in order to implement a decorator.

I'd be very curious to know more about this. Do you have an example that you could share?

So if I understand correctly, in Sangria you use FieldTags when GraphQL-JS uses schema directives (except FieldTags are only for fields), and you use middleware to run some logic based on the metadata (which is still missing in GraphQL-JS, but would be present in decorators)?

> You can't really cover this use-case with decorators because you need to manually decorate every single field and type in the schema

I think I did not explain it very well, but decorators as I envisioned them could be applied to types and to the whole schema as well. So you could have a decorator on the schema, which modifies every resolve function in the entire schema, essentially making it a middleware. This could also do the caching and logging that you mentioned.

With some contortions, I believe decorators could work for all the other cases that you mentioned as well, but it feels much less natural, so I'm starting to warm up to the idea of middleware. 

However, there are still some things that are not possible with FieldTags and middleware, such as modifying the schema, i.e. adding things such as `@deprecationReason` and `@description` or creating new fields or arguments based on decorators (not saying all of those are good ideas, but there may be legitimate reasons to use them).

> It would help if validation logic is performed even before query is executed
I agree 100% here. I think the concept of query reducers is very useful, and decorators don't help with that (although you could essentially create the equivalent of a reducer with a decorator that wraps the whole schema.

I think most of the decorators proposal could be implemented with annotations and middleware (or schema pre-processors). I'd be interested in exploring that direction and making sure that these annotations, middleware, reducers and pre-processors are compatible across different implementations of GraphQL, such as GraphQL-JS, Sangria, Graphene and others. Would you guys be interested in working on this together @OlegIlyenko @syrusakbary?

I'll take another look at Sangria middleware and reducers and try to rework this proposal in a way that works well with them and also incorporates the proposed schema directives of GraphQL-JS. If you have any ideas, please share!

**Comment by OlegIlyenko** [comment]
> I'd be very curious to know more about this. Do you have an example that you could share?

I guess the terminology gets a bit confusing. I think all examples I provided demonstrate it in some form.

> So if I understand correctly, in Sangria you use FieldTags when GraphQL-JS uses schema directives (except FieldTags are only for fields), and you use middleware to run some logic based on the metadata (which is still missing in GraphQL-JS, but would be present in decorators)?

For the moment being, directives exist only in form of pull requests. So I haven't though much about it yet. After looking at the way they are defined now, I think it makes a lot or sense to replace field tags with the IDL directives. I added `FieldTag` quite a while ago, even before this whole annotation/directive discussion started, so I think now comes the point where I need to think how to unify these concepts.

One thing I would like to point out though, is that I still see IDL as a secondary in comparison to a schema definition in the code itself (e.g in javascript with `GraphQLObjectType`). I definitely see some use-cases for an pure IDL schema definition, that's why I also excited about directive addition. They will allow to customize particular aspects of schema generation. I just try to be cautious when it comes to things that one actually expresses with the directives in the IDL. For example one can take it to an extreme and define the whole database ORM framework around it:

```js
type Query {
  users: [User] @SQLQuery("select * from users")
}
```

or something along these lines. I definitely see value in this, but I would approach development of such framework very carefully.  It may open a Pandora's box. I find it much easier sometimes to just write 2 lines of simple javascript code than to bend the framework to your will for days :) 

So this generally means that I first try to think how this concept would be represented in code itself (when you define schema in the code) and then whether it makes sense to expose it as a directive in the IDL. 

Sorry, it a bit of random train of though inspired by the decorators proposal and this discussion :) I guess it's a bit too abstract to serve any practical purpose :)

> However, there are still some things that are not possible with FieldTags and middleware, such as modifying the schema, i.e. adding things such as @deprecationReason and @description or creating new fields or arguments based on decorators (not saying all of those are good ideas, but there may be legitimate reasons to use them).

Indeed, you are right - it's not possible at the moment. But in the other hand it's by intention. In theory it should be possible to implement this functionality, but I would really love to see the actual real-word use-case for it. I think the primitives that graphql-js as well as any other GraphQL library provides are very suitable for most of use-cases. If one need to generate parts of the schema, than it's pretty straightforward to just write a function that does this. graphql-relay-js is a good example of it. And this is something that would be it easy to understand for most of the users. Introducing an annotation/directive/decorator based framework for schema generation maybe be suitable for some use-cases, but it has a learning curve associated with it and increases conceptual surface area of the library. I'm not saying that it's not worth doing though, I guess I just being a bit too cautious when it comes to this kind of stuff :)

I was thinking a bit more about it. If we stick to a term "decorator", maybe introduction of following concepts may help to cover use-cases we described above (and make it easier to reason about):

* **decorator annotation** - it provides a piece of meta-data and can be mapped to an IDL directive. It doesn't do anything by itself though, it always requires the second component: decorator runtime. Every decorator annotation always have semantics associated with it (which is documented in some form).
* **decorator runtime** - This one encapsulates the run-time behavior of the decorator. Decorator annotation may have some code and logic in it, but the decorator runtime is the component that is responsible to run/execute this logic at particular moment in time during the query execution or query analysis. Decorator runtime itself has 2 phases (or it can be also just split in 2 components)
  * **query analysis phase** - at this phase decorator runtime may analyze the query with the help of decorator annotations and do some interesting stuff with this analyzed data (validate arguments, analyze complexity, go to database to fetch additional information, etc.). At this point query execution is not started yet, which means that decorator runtime is able stop the execution if necessary.
  * **query execution phase** - that this phase decorator is able to participate in query execution and modify the runtime behavior of fields and types is some ways (metrics, authorization, caching, etc.)

This means that there should be a way not only to define directives in IDL, but also to provide a list of default directive runtimes (which also can be inferred based on used directives).

One aspect of it I find particularity important about this idea: at no point GraphQL schema is modified or mutated in any way (in terms of javascript object mutation). This means that it's possible to customize the list of decorator runtimes for every GraphQL query execution:

```js
executeGraphQLQuery({query: query, schema: schema, decoratorRuntimes: [validateArgs, reportMetrics, authorizeFields]}})
```

I think this can be a very flexible way to work with decorators.

At the end of the day, on a conceptual level for end-user `@validateRange(from: 0, to: 10)` can be seen as a decorator that "wraps" a resolve function. The fact that it's separated in 2 components in order to optimize it's runtime behavior can be seen as just an implementation detail. 

Let me know what you think about this :) It's just a barnstorming idea

**Comment by dcworldwide** [comment]
Hugely powerful concept. Excited to use it. Just wondering why not keep with the pseudo-standard @decorator paradigm?

**Comment by OlegIlyenko** [comment]
After thinking a bit more about this, it may be also helpful to introduce a third phase to the decorator runtime:

* **query validation phase** - at this phase decorator runtime participates in query validation provided by current implementation of graphql-js. It can be done in form of a validation rule. This will make it easier for a decorator to produce validation errors in a standard format.

**Comment by stubailo** [comment]
@dcworldwide it was initially to not conflict with GraphQL directives, but it looks like those might now be a part of the implementation. I think @helfer is working on an updated design with that in mind.

@OlegIlyenko it would be absolutely spectacular if it were possible to add more restrictive validation via decorators, and then use it in tools like https://github.com/apollostack/eslint-plugin-graphql

**Comment by OlegIlyenko** [comment]
@stubailo I haven't thought about it, but indeed, by integrating decorators with existing validation mechanism we also get other benefits, like generic tooling support.

**Comment by helfer** [comment]
@dcworldwide In this proposal I didn't use `@` because GraphQL has so-called directives which can be applied to a query. They use the `@` notation, but they come **after** the thing they modify. Because of the significant potential for confusion, and because decorators in virtually every other language come **before** the thing they decorate, I decided to propose putting them in front, but use a different symbol.

However, based on the recent PR adding directives to the schema language as well, I think I'll rewrite this proposal to make it based on directives. That means we'll probably have decorators that come after the thing they decorate, which is an okay compromise to keep consistency between the query language and the schema language.

**Comment by helfer** [comment]
@OlegIlyenko 

> I just try to be cautious when it comes to things that one actually expresses with the directives in the IDL. For example one can take it to an extreme and define the whole database ORM framework around it.

I definitely agree, it doesn't make sense to try forcing everything into a DSL when defining it in code works just as well. In my opinion, the main benefit of using the IDL is clarity and legibility. For things that can be done declaratively, it makes sense to do so. Just like it is easier to write declarative data requirements of a component with GraphQL than it is to write some runtime data fetching logic as we currently do for most REST endpoints.

One of the coolest things about declarative information and the IDL is that it opens the possibilities for static analysis tools that can be used across all GraphQL implementations in the form of generic tools (as @stubailo and you already hinted at in the context of validators). For example, if we have a standard role-based permission system, one can declaratively specify who can see what, and it would be possible to use tools like [Alloy](http://alloy.mit.edu/alloy/) to find possible counter-examples where the security rules can be broken.

> This means that there should be a way not only to define directives in IDL, but also to provide a list of default directive runtimes (which also can be inferred based on used directives).

Yes, essentially that's already what my proposal did, but in the code the things were not that separate. Maybe it makes sense to separate the two concepts so it's clear that they can be used independently. I think it still makes sense to ship the directive definition and the runtime code in one package, but there should be a way to use them independently.

> At the end of the day, on a conceptual level for end-user @validateRange(from: 0, to: 10) can be seen as a decorator that "wraps" a resolve function. The fact that it's separated in 2 components in order to optimize it's runtime behavior can be seen as just an implementation detail.

That makes sense. I wrote the paragraph above just before reading this 👍 

> I think this can be a very flexible way to work with decorators (this does not mean that you can't implement thins like @description or @addField - a minor changes in execution engine may consider decorators in introspection and field collection).

I like the idea of having different phases as you proposed. I think with the addition of a "schema postprocessing" phase, decorators that modify the schema could be implemented as well. A good reason for having explicit phases is also that it could make it easy for developers to know what a certain decorator does/doesn't do. If the decorator says that it doesn't run during schema creation, then they know it doesn't modify the objects. If it only runs during validation, they know it won't affect execution: the query will either fail completely, or be executed, but not fail somewhere in the middle.

I read the documentation about sangria middleware and query reducers. I will incorporate all of that information and your great feedback into the next version of this proposal! Let's keep up the conversation!

**Comment by tlvenn** [comment]
Hi @helfer , I was wondering if you have been able to work on the next version of this proposal ? Kinda excited to see where this is going. Thanks a lot for leading this effort and opening the discussion around it.

**Comment by helfer** [comment]
@tlvenn Thanks, I had to work on a bunch of other stuff the last few weeks, but it looks like I should be able to focus on this again next week, so stay tuned!

**Comment by krugar** [comment]
this may be a bit of a tangent, as it is more of a use case than a feature, but i think this might be a good place to leave this thought anyways:

i'd like to see a "+context"-decorator that facilitates the mapping of a (part of) a graphQL-js schema to one or more schema.org or opengraph schemas, nudging users towards standardized data types/namespaces (the syntactic details would need some fleshing out in the case of multiple contexts).

usually, the types defined by schema.org do not exactly fit the local requirements, so you end up with lots of unused/empty schema.org fields and some additional fields that either have no formal description or that you don't want to hard-link to a formally described meaning for some reason.

but still, if there was tooling around mapping e.g. the ubiqutous "User" object to http://schema.org/Person, then one could dream of a future with a growing ecosystem of reusable frontend components that are aligned to common data-type standards, not only not missing display options for some of the now-expectable fields, but also marking them up in RDFa or JSON-LD for full semantic web goodness (and SEO benefits, i guess).

like i said, tangent. but since you guys are aiming to make apollo the gold standard of data loading, maybe it would be a good place to promote (semantic web) data standards in some way as well?

**Comment by helfer** [comment]
@krugar Yeah, that's definitely one possible use-case. I think the mapping should be between GraphQL types and schema.org types.

**Comment by antmdvs** [comment]
It would interesting to see this gain traction. Are you waiting to see how GraphQL evolves?

**Comment by helfer** [comment]
@antmdvs I'm still interested in the idea, but I didn't have time to work on it. If I have some spare cycles, I might pick it up. If anyone else wants to give it a shot, I'd be happy to help as well!

**Comment by almilo** [comment]
@helfer I have been working on this concept here under the name of *annotations* instead of *decorators* and with the *directives* syntax: https://github.com/almilo/granate.
The more I advance, the more I see part of it as an extension of the graphql-tools. I have so far implemented some infrastructure to extract the *annotations* from the schema language, a wrapper for express-graphql, graphql and graphql-tools and two *annotations*: *mock* and  *rest*.
I will be off a week but after that I plan to work all October on expanding the concept. What do you think?

**Comment by helfer** [comment]
@almilo That looks really interesting, and I really like the mock and rest decorators.

However, I'm concerned that people will be much less likely to use it if it involves a CLI that has to be invoked every time something changes. I think making each of the parts (annotation extractor, etc.) available separately would make it much more likely that other people use your code.

**Comment by almilo** [comment]
@helfer using it from the CLI has a clear use case IMO and that is rapid API prototyping / mocking / migration of existing REST APIs, etc. But I'm also aware of your concern and that is why I created a separate granate-cli package. What I'm wondering is how to re-distribute the implementation. At the moment, granate is mostly a wrapper around graphql-tools and I'm even thinking of dropping the usage of graphql buildSchema + rootValue approach of 0.7.0 in favour of pure graphql-tools resolvers. If done so, then applyDecorators could be a further functionality of graphql-tools, don't you think? The extractor from the schema language and the two decorators mock and rest could be moved to this repo and granate would be just focused on the execution of decorated schemas (I have some clear use cases for that). I would even move the CLI commands to granate from granate-cli.
The naming (annotations/decorators) can be discussed. Does this make sense to you in order to bring the decorators concept forward?

**Comment by helfer** [comment]
@almilo sorry I didn't get back to you on this. Unfortunately I don't have time to work on it at the moment, but if you ping me again at the end of next week, things should be looking better 🙂 

I think the first steps in moving decorators forward is to check if the directives as they are implemented now would actually support the use of decorators. To do that I would just create one or two decorators, and then write an `applyDecorators` function as you proposed. I assume there will be a few blockers on the way, which will need PRs to graphql-js, so it could be useful to do that now.

**Comment by voodooattack** [comment]
Any news on this? 

I have a use case in my new project (unifying GraphQL type definitions and database models), and I don't want to write a custom schema language parser just yet.

**Comment by helfer** [comment]
@voodooattack sadly still not. But maybe we'll get some people working on this in January as part of a coordinated effort to do more exploratory projects like this.

**Comment by anton-matosov** [comment]
Hi @helfer, I want to give decorators a try and contribute firebase `connector` decorator. I also want to try to implement firebase rules generation using graphql schema and decorators. It looks like a perfect fit for these needs.
Can you please point out where I can get latest source for connectors and decorators?

**Comment by nunofgs** [comment]
Any news on decorators? My use case is doing client-side validation as well.

**Comment by mellocello2003** [comment]
I'd love to see decorators available as well.

**Comment by codepunkt** [comment]
Same here. How can we help?

**Comment by timrs2998** [comment]
Same, we're also looking for news on decorators and any updates would be awesome

**Comment by falconmick** [comment]
I'll join the list of waiting for info!

**Comment by ThisIsMissEm** [comment]
Don't graphql directives now supersede this project? https://dev-blog.apollodata.com/reusable-graphql-schema-directives-131fb3a177d1

**Comment by b3ross** [comment]
@ThisIsMissEm thanks for the info!

**Comment by taylorgoolsby** [comment]
@voodooattack 
> Any news on this?
> 
> I have a use case in my new project (unifying GraphQL type definitions and database models), and I don't want to write a custom schema language parser just yet.

I have a similar use case. I want to write a schema in GraphQL SDL, and then decorate it so that I process the augmented `schema.json` by another parsing tool which will generate a `.sql` script. Has anyone done open sourced this yet?

**Comment by enumag** [comment]
@taylorgoolsby Converting GraphQL schema to SQL schema? I have been looking for that for a long time. Are you planning to open-source it?

**Comment by taylorgoolsby** [comment]
@enumag @voodooattack 
After looking at ThisIsMissEm's referal to gql directives, I think that's the better option than decorators. 

Thank you for the referal to TypeORM, voodooattack. That looks really interesting as a way of generalizing the solution. I'm very interested in an implementation of something that uses GraphQL Directives to selectively port a SDL schema to a bunch of TypeORM entities. Something like this would be great:
`yarn add graphqlorm`
```
import {createConnection} from "graphqlorm";

createConnection({
    type: "mysql",
    host: "localhost",
    port: 3306,
    username: "root",
    password: "admin",
    database: "test",
    makeExecutableSchemaOptions,
    synchronize: true,
}).then(connection => {
    // here you can start to work with your entities
}).catch(error => console.log(error));
```
 
So basically, is a combination of the API from `typeorm` and `makeExecutableSchema`.

Maybe the directives look like this in the SDL schema:
```
type User @orm(options: CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci) {
  userId: String! @orm(primary: true, type: BINARY(16) NOT NULL)
  databaseOnlyField: Int! @orm(databaseOnly: true)
  graphqlOnlyField: String!
  posts: [Post!]! @orm(oneToMany: true, relayConnection: true)
}

type Post @orm {
  postId: Int! @orm(primary: true, autoincrement: true)
  content: String! @orm(length: 240, unicode: true)
  dateCreated: Date! @orm
}
```

I'm don't know much about TypeORM, so I can't say what the best annotation arguments should be to work with TypeORM. Ideally it would be something that makes sense to people who already know TypeORM, and also requires the least amount of code to translate into TypeORM, but at the same time, it should leave the SDL schema looking as neat as possible.

How do you think it will take to port that over to GraphQL Directives voodooattack? I need something like this soon. Originally, I was thinking of something that would output a `.sql` script in MySQL from an SDL schema. For me, that's a easier difficulty task since I don't know TypeORM at the moment. I might have to make a MySQL only version, just to accomplish my needs soon, but I'll looking forward to your TypeORM version.

**Comment by voodooattack** [comment]
@taylorgoolsby @enumag 

You guys are in luck. Guess what I just found by digging around a bit? 

https://gist.github.com/voodooattack/0ac624fcc9e33cf807ce209937ff9bd2

This is an old (and quite hacky) implementation I found in a gist that I must have extracted at one point from a previous project that used Sequelize. It's about a year old, so it could be outdated, and it seems to be missing some features, but the most important part is this: it generates an entire relational database and the resolvers to manipulate it for you; and it handles access control to individual resolvers as well.

Might be a good reference (and a good starting point) for how this could work with TypeORM though, since it would be easier to work with it! (from personal experience, as I remember rewriting this to use TypeORM at one point)

Unfortunately, I don't have access to the repo and/or the latest sources for that project anymore. As I finished and turned in that contract early this year.

**Comment by taylorgoolsby** [comment]
@voodooattack @enumag 

Prisma might also be an option: https://www.prisma.io/docs/reference/upgrade-guides/graphcool-to-prisma/data-modelling-and-graphql-api-eiwuezeec3/

**Comment by voodooattack** [comment]
> Prisma might also be an option: https://www.prisma.io/docs/reference/upgrade-guides/graphcool-to-prisma/data-modelling-and-graphql-api-eiwuezeec3/

This looks interesting! Last time I checked something like this, all available ones were SaaS-only solutions and didn't allow you to host your own server/database.

I guess I don't have to work on this then. Prisma looks more suitable! :+1:

**Comment by TimMikeladze** [comment]
> Prisma seems heavily invasive though. Correct me if I'm wrong, but it looks like an all or nothing approach to using ORMs. I would like the ability to opt out of using the ORM and use something like `find_by_sql` if ncessary. It feels like if you use Prisma, you have to buy into the whole Prisma ecosystem.

When using prisma you choose which resolvers you want to forward to the prisma data layer, that means you can define your own types and resolvers which are in no way dependent on prisma.

**Comment by enumag** [comment]
I looked into prisma before but I don't think it's what I want. It seems that it generates most of the final GraphQL schema automatically. I don't want that, I want to design all of the schema myself, but I don't really care what the underlying database looks like - I can feed the data into whatever structure needed because I'm using CQRS and event store.

**Comment by taylorgoolsby** [comment]
@enumag 
To calrify what your's saying about what you don't like about Prisma, are you saying you don't like this feature: https://www.prisma.io/docs/data-model-and-migrations/data-model-knul/#datamodel-vs-prisma-graphql-schema
To summarize, the part of Prisma you have control over, the datamodel, which is defined using GraphQL SDL, is not actually a GraphQL Schema, and under-the-hood, Prisma converts this datamodel to the actual GraphQL schema, and there is no way to override or modify this under-the-hood part.

I agree even though I'm not using CQRS. In my case, I don't have a CRUD API, and instead, I want anemic mutations: https://medium.com/@__xuorig__/graphql-mutation-design-anemic-mutations-dd107ba70496

**Comment by taylorgoolsby** [comment]
@enumag Please, try out https://www.npmjs.com/package/graphql-to-sql
